can/zynqmp-can/zynqmp-can.c

```diff
@@ -0,0 +1,565 @@
+				can.base->ier &= ~(1 << 7);
+				fprintf(stderr, "zynqmp-can: unknown condWait error %i\n", condRet);
+				return condRet;
+			}
+		}
+	}
+
+	/* Read all received frames */
+	for (uint32_t i = 0; i < bufLen; i++) {
+		/* Try clearing the interrupt (it will only clear if there is no data) */
+		can.base->icr = (1 << 7);
+
+		/* Check if there is still data in the RX FIFO (by reading pending IRQ status) */
+		if (0 == (can.base->isr & (1 << 7))) {
+			break;
+		}
+
+		/* Read a frame from the FIFO */
+		buf[i].id = (can.base->rxfifo_id >> 21) & 0x7FF;
+		buf[i].len = (can.base->rxfifo_dlc >> 28) & 0xF;
+		buf[i].payload.words[0] = be32toh(can.base->rxfifo_data1);
+		buf[i].payload.words[1] = be32toh(can.base->rxfifo_data2);
+		(*recvFrames)++;
+	}
+
+	return EOK;
+}
+
+/**
+ * Main thread for handling CAN operations.
+ * Processes incoming messages and performs the requested operations.
+ */
+static void zynqmpCan_thread(void *arg)
+{
+	msg_t msg;
+	msg_rid_t rid;
+	uint32_t port = can.oid.port;
+
+	mutexLock(can.lock);
+
+	for (;;) {
+		/* Wait indefinitely for a message */
+		if (msgRecv(port, &msg, &rid) < 0) {
+			continue;
+		}
+		switch (msg.type) {
+			/* This driver supports only one way of communication with it */
+			case mtDevCtl: {
+				/* Decode extra information passed with the request */
+				zynqmp_canDriverReq *req = (zynqmp_canDriverReq *)msg.i.raw;
+				zynqmp_canDriverResp *resp = (zynqmp_canDriverResp *)msg.o.raw;
+
+				/* Choose operation based on opcode */
+				if (req->operation == zynqmp_canOpSend) {
+					zynqmp_canFrame *frames = (zynqmp_canFrame *)msg.i.data;
+					uint32_t count = msg.i.size / sizeof(zynqmp_canFrame);
+					msg.o.err = zynqmpCan_sendFrames(frames, count, req->blockTxOperation, &resp->framesReceivedSend);
+				}
+				else if (req->operation == zynqmp_canOpRecv) {
+					zynqmp_canFrame *buf = (zynqmp_canFrame *)msg.o.data;
+					uint32_t bufLen = msg.o.size / sizeof(zynqmp_canFrame);
+					msg.o.err = zynqmpCan_recvFrames(buf, bufLen, req->timeoutUs, &resp->framesReceivedSend);
+				}
+				else {
+					fprintf(stderr, "zynqmp-can: unknown opcode\n");
+					msg.o.err = -EINVAL;
+				}
+			} break;
+
+			case mtOpen:
+			case mtClose:
+				msg.o.err = EOK;
+				break;
+
+			case mtWrite:
+			case mtRead:
+			case mtGetAttr:
+			default: {
+				msg.o.err = -ENOSYS;
+				break;
+			}
+		}
+		msgRespond(port, &msg, rid);
+	}
+
+	mutexUnlock(can.lock);
+}
+
+/**
+ * Main entry point for the CAN driver.
+ * Parses command-line arguments, initializes the CAN peripheral, and starts the message handling thread.
+ */
+int main(int argc, char **argv)
+{
+	int c = 0;
+
+	/* Default arguments */
+	int can_id = can_1;
+	int baudrate = CAN_BAUDRATE_KBPS_DEFAULT;
+
+	/* Parse command-line arguments */
+	if (argc > 1) {
+		while ((c = getopt(argc, argv, "n:b:h")) != -1) {
+			switch (c) {
+				case 'b': {
+					baudrate = atoi(optarg);
+				} break;
+
+				case 'n': {
+					can_id = atoi(optarg);
```

Comment:
According to MISRA - avoid casting signed -> enum.

IMHO `can_0`, `can_1` enums serve little purpose, plain int suffices (we store the peripheral number in variable called `can_id`, it's plenty descriptive)